File: web_sdk/src/client/otpk-replenisher.ts
// src/client/otpk-replenisher.ts

import { KeyGenerator } from '../crypto/key-generator.js';
import {
  exportPublicKeyRaw,
  exportPrivateKeyPkcs8,
} from '../crypto/crypto-utils.js';
import { base64urlEncode } from '../utils/encoding.js';
import type { SessionStore } from '../session/session-store.js';
import type { RestClient } from '../transport/rest-client.js';
import type { Logger } from '../utils/logger.js';
import type { SdkError } from '../errors/sdk-error.js';
import type { TypedEventEmitter } from '../utils/typed-emitter.js';

const REPLENISH_COUNT = 50;
const LOW_OTPK_THRESHOLD = 10;

/**
 * Listens for low_otpk events and automatically replenishes one-time prekeys.
 */
export class OtpkReplenisher {
  constructor(
    private readonly userId: string,
    private readonly deviceId: string,
    private readonly store: SessionStore,
    private readonly rest: RestClient,
    private readonly logger: Logger,
    private readonly onError: (err: SdkError) => void,
  ) {}

  async handleLowOtpk(count: number): Promise<void> {
    if (count >= LOW_OTPK_THRESHOLD) return;

    this.logger.info(`Low OTPK count (${count}), replenishing ${REPLENISH_COUNT}`);

    try {
      const deviceRecord = await this.store.loadDevice(this.userId, this.deviceId);
      if (!deviceRecord) return;

      const newOtpks = await KeyGenerator.generateOtpks(REPLENISH_COUNT, deviceRecord.nextOtpkId);

      // Build public key entries for server
      const entries = await Promise.all(
        newOtpks.map(async (o) => ({
          id: o.id,
          key: base64urlEncode(await exportPublicKeyRaw(o.keyPair.publicKey)),
        })),
      );

      await this.rest.replenishOtpks(this.userId, this.deviceId, entries);

      // Persist new private OTPKs
      const newPrivateOtpks = await Promise.all(
        newOtpks.map(async (o) => ({
          id: o.id,
          privateKey: base64urlEncode(await exportPrivateKeyPkcs8(o.keyPair.privateKey)),
          publicKey: base64urlEncode(await exportPublicKeyRaw(o.keyPair.publicKey)),
        })),
      );

      await this.store.saveDevice({
        ...deviceRecord,
        otpks: [...deviceRecord.otpks, ...newPrivateOtpks],
        nextOtpkId: deviceRecord.nextOtpkId + REPLENISH_COUNT,
      });

      this.logger.info(`Replenished ${REPLENISH_COUNT} OTPKs`);
    } catch (err) {
      this.logger.error('OTPK replenishment failed', err);
      this.onError(err as SdkError);
    }
  }
}

File: flutter_sdk/lib/src/client/otpk_replenisher.dart

import '../transport/rt_event.dart';
import '../transport/rest_client.dart';
import '../session/session_store.dart';
import '../session/session_data.dart';
import '../crypto/key_generator.dart';
import '../crypto/crypto_types.dart';
import '../utils/logger.dart';

/// Automatically replenishes one-time pre-keys when the server reports
/// a [LowOtpkRtEvent] with count < 10.
///
/// Generates 50 fresh OTPKs, uploads them, and updates the [DeviceRecord]
/// in the [SessionStore].
class OtpkReplenisher {
  final RestClient _restClient;
  final SessionStore _sessionStore;
  final String _userId;
  final String _deviceId;
  final SdkLogger _logger;

  static const _replenishCount = 50;
  static const _lowThreshold = 10;

  OtpkReplenisher({
    required RestClient restClient,
    required SessionStore sessionStore,
    required String userId,
    required String deviceId,
    SdkLogger? logger,
  })  : _restClient = restClient,
        _sessionStore = sessionStore,
        _userId = userId,
        _deviceId = deviceId,
        _logger = logger ?? SdkLogger(tag: 'OtpkReplenisher');

  /// Called by [RealTimeEventRouter] when a [LowOtpkRtEvent] is received.
  Future<void> onLowOtpk(LowOtpkRtEvent event) async {
    if (event.count >= _lowThreshold) return;
    _logger.info(
        'Low OTPK warning: ${event.count} remaining. Replenishing $_replenishCount keys...');

    try {
      final newKeys = await KeyGenerator.generateOneTimePreKeys(_replenishCount);

      final publicKeys = newKeys
          .map((k) => OneTimePreKey(
                id: k.id,
                publicKey: k.keyPair.publicKey,
              ))
          .toList();

      await _restClient.replenishOtpks(_userId, _deviceId, publicKeys);

      // Update stored DeviceRecord with new private key material
      final record = await _sessionStore.loadDevice(_userId, _deviceId);
      if (record != null) {
        final newOtpkRecords = newKeys
            .map((k) => OtpkRecord(
                  id: k.id,
                  privateKey: k.keyPair.privateKey,
                  publicKey: k.keyPair.publicKey,
                ))
            .toList();

        await _sessionStore.saveDevice(
          record.copyWith(
            oneTimePrekeys: [...record.oneTimePrekeys, ...newOtpkRecords],
          ),
        );
      }

      _logger.info('Successfully replenished $_replenishCount OTPKs');
    } catch (e) {
      _logger.error('Failed to replenish OTPKs: $e');
    }
  }
}
